src/storage.py

```diff
@@ -0,0 +1,267 @@
+        """Updates the attached storage devices in state."""
+        # check if the wal/db device is being used.
+        osd_num = self._get_osd_num(event.storage.location.as_posix(), event.storage.name)
+
+        # If detaching wal/db disk is used with an OSD.
+        if osd_num:
+            with guard(self.charm, self.name):
+                try:
+                    self.remove_osd(osd_num)
+                except CalledProcessError as e:
+                    if self._is_safety_failure(e.stderr):
+                        warning = f"Storage {event.storage.full_id} detached, provide replacement for osd.{osd_num}."
+                        logger.warning(warning)
+                        # clean records since juju will deprovision device.
+                        self.remove_osd(osd_num, force=True)
+                        raise BlockedExceptionError(warning)
+
+    """helpers"""
+
+    def _is_safety_failure(self, err: str) -> bool:
+        """Checks if the subprocess error is caused by safety check."""
+        return "need at least 3 OSDs" in err
+
+    def _run(self, cmd: list) -> str:
+        """Wrapper around subprocess run for storage commands."""
+        process = run(cmd, capture_output=True, text=True, check=True, timeout=180)
+        logger.debug(f"Command {' '.join(cmd)} finished; Output: {process.stdout}")
+        return process.stdout
+
+    def _enroll_with_wal_db(self, disk: list, wal: list, db: list):
+        """Checks if sufficient devices are available to be enrolled into OSDs."""
+        enrollment_count = min(len(disk), len(wal), len(db))
+        for i in range(enrollment_count):
+            try:
+                microceph.add_osd_cmd(disk[i], wal[i], db[i])
+                # store configured devices.
+                self._save_osd_data(disk=disk[i], wal=wal[i], db=db[i])
+            except CalledProcessError as e:
+                logger.error(e.stderr)
+
+    def _enroll_disks_in_batch(self, disks: list):
+        """Adds requested Disks to Microceph and stored state."""
+        microceph.enroll_disks_as_osds(disks)
+        for disk in disks:
+            self._save_osd_data(disk)
+        logger.debug(f"Added {disks} as OSDs.")
+
+    def remove_osd(self, osd: int, force: bool = False):
+        """Removes OSD from MicroCeph and from stored state."""
+        try:
+            microceph.remove_disk_cmd(osd, force)
+            self._remove_osd_data(self._stored.osd_data[osd])
+        except CalledProcessError as e:
+            if force:
+                # clean stored state.
+                self._remove_osd_data(self._stored.osd_data[osd])
+            raise e
+
+    def _save_osd_data(self, disk: str, wal: str = None, db: str = None):
+        """Save OSD data."""
+        for osd in microceph.list_disk_cmd()["ConfiguredDisks"]:
+            # e.g. check 'vdd' in '/dev/vdd'
+            if str(microceph._get_disk_info(osd["path"], "name")) in disk:
```

Comment:
The osd["path"] value for a block device would be the `/dev/disk-by-id/` value which should be unique for each configured disk. We check this value in `_get_disk_info` which is a wrapper around `lsblk --json`. Thus this would not give false positive matches.